WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,199 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <algorithm>
+
+#include <tilegen/CartesianPower.h>
+
+class TestCartesianPower : public ::testing::Test
+{
+protected:
+    typedef std::vector<int> Container;
+    typedef std::vector<int>::const_iterator Iterator;
+
+    Container base{4, 5, 6};
+    const tilegen::CartesianPower<Iterator> cp{base, 4};
+};
+
+TEST_F(TestCartesianPower, TestCartesianPowerSize)
+{
+    EXPECT_EQ(81, cp.size()) <<
+        "CartesianProduct has incorrect size";
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCount)
+{
+    size_t count = 0;
+    for (const auto& it : cp)
+        count++;
+    EXPECT_EQ(81, count);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonBeginEnd)
+{
+    EXPECT_EQ(cp.cbegin(), cp.cbegin());
+    EXPECT_NE(cp.cend(), cp.cbegin());
+    EXPECT_NE(cp.cbegin(), cp.cend());
+    EXPECT_EQ(cp.cend(), cp.cend());
+
+    EXPECT_EQ(cp.cbegin(), cp.begin());
+    EXPECT_EQ(cp.cend(), cp.end());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonLast)
+{
+    auto it = cp.cbegin();
+    for (size_t i = 0; i < cp.size(); i++)
+        ++it;
```

Comment:
Could you use `std::advance` here? It will be nice, if it works the same way as standard iterators.